firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigAutoFetch.java

```diff
@@ -0,0 +1,175 @@
+
+  private static final int FETCH_RETRY = 3;
+
+  private final Set<ConfigUpdateListener> eventListener;
+
+  private final HttpURLConnection httpURLConnection;
+  private final ConfigFetchHandler configFetchHandler;
+  private final ConfigUpdateListener retryCallback;
+  private final ScheduledExecutorService scheduledExecutorService;
+  private final Random random;
+  private final Executor executor;
+
+  public ConfigAutoFetch(
+      HttpURLConnection httpURLConnection,
+      ConfigFetchHandler configFetchHandler,
+      Set<ConfigUpdateListener> eventListener,
+      ConfigUpdateListener retryCallback,
+      Executor executor,
+      ScheduledExecutorService scheduledExecutorService) {
+    this.httpURLConnection = httpURLConnection;
+    this.configFetchHandler = configFetchHandler;
+    this.eventListener = eventListener;
+    this.retryCallback = retryCallback;
+    this.scheduledExecutorService = scheduledExecutorService;
+    this.random = new Random();
+    this.executor = executor;
+  }
+
+  public void beginAutoFetch() {
+    executor.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            listenForNotifications();
+          }
+        });
+  }
+
+  // Check connection and establish InputStream
+  private void listenForNotifications() {
+    if (httpURLConnection != null) {
+      try {
+        int responseCode = httpURLConnection.getResponseCode();
+        if (responseCode == 200) {
+          InputStream inputStream = httpURLConnection.getInputStream();
+          handleNotifications(inputStream);
+          inputStream.close();
+        } else {
+          for (ConfigUpdateListener listener : eventListener) {
+            listener.onError(
+                new FirebaseRemoteConfigRealtimeUpdateStreamException(
+                    "Http connection responded with error: " + responseCode));
+          }
+        }
+      } catch (IOException ex) {
+        for (ConfigUpdateListener listener : eventListener) {
+          listener.onError(
+              new FirebaseRemoteConfigRealtimeUpdateFetchException(
+                  "Error handling stream messages while fetching.", ex.getCause()));
+        }
+      }
+    }
+    retryCallback.onEvent();
+  }
+
+  // Auto-fetch new config and execute callbacks on each new message
+  private void handleNotifications(InputStream inputStream) throws IOException {
+    BufferedReader reader = new BufferedReader((new InputStreamReader(inputStream, "utf-8")));
+    String message;
+    while ((message = reader.readLine()) != null) {
+      long targetTemplateVersion = configFetchHandler.getTemplateVersionNumber();
+      try {
+        JSONObject jsonObject = new JSONObject(message);
+        if (jsonObject.has("latestTemplateVersionNumber")) {
+          targetTemplateVersion = jsonObject.getLong("latestTemplateVersionNumber");
+        }
+      } catch (JSONException ex) {
+        Log.i(TAG, "Can't get latest config update message.");
```

Comment:
To be strict, you could get the config, but you couldn't parse it.

Reply:
Updated comment to include parsing error